torch/onnx/_internal/fx/passes/type_promotion.py

```diff
@@ -1469,40 +1469,68 @@ def _maybe_promote_arg(
 
         if isinstance(fx_arg, torch.fx.Node):
             arg_val = self.env[fx_arg]
-            if (old_dtype := arg_val.dtype) != dtype:
-                # Promote tensor to dtype.
-                graph = node.graph
-                with graph.inserting_before(node):
-                    diagnostic.with_additional_message(
-                        f"Argument {fx_arg}({old_dtype}) is promoted to {dtype}."
-                    )
-                    return self._create_node(
-                        graph,
-                        "call_function",
-                        torch.ops.prims.convert_element_type.default,
-                        (fx_arg,),
-                        {"dtype": dtype},
-                    )
-            diagnostic.with_additional_message(
-                f"Argument {fx_arg} is not promoted. Already {dtype}."
-            )
-            return fx_arg
+            if isinstance(arg_val, torch.Tensor):
+                if (old_dtype := arg_val.dtype) != dtype:
+                    # Promote tensor to dtype.
+                    graph = node.graph
+                    with graph.inserting_before(node):
+                        diagnostic.with_additional_message(
+                            f"Argument {fx_arg}({old_dtype}) is promoted to {dtype}."
+                        )
+                        return self._create_node(
+                            graph,
+                            "call_function",
+                            torch.ops.prims.convert_element_type.default,
+                            (fx_arg,),
+                            {"dtype": dtype},
+                        )
+                diagnostic.with_additional_message(
+                    f"Argument {fx_arg} is not promoted. Already {dtype}."
+                )
+                return fx_arg
+            elif isinstance(arg_val, proxy_tensor.py_sym_types):
```

Comment:
Not related, but what we can think about.

Can't help but notice we have `if isinstance(value, torch.Tensor) ... elif  isinstance(value, sym_py) ...`. Maybe there is a way we can abstract this if/else branch, and simplify the code everywhere. hmmmm

Reply:
I almost wanted to do a factory + visitor pattern rewrite this time. However, each visitor class needs access to interpreter details in order to create new nodes and run. I might work on it if we see more cases.